### ISSUE TITLE
Growth rate not fitted correctly across countries

### ISSUE THREAD

**Issue by thiemonagel** [nicebread/corona#6]
Afaics the overall growth rate is determined by fitting intercept*(1+b)^day to all data points (across countries). Instead, I believe that the intercept should vary by country and only b should be universal across countries.

See this example:
US growth rate is fitted to 31% here: https://imgur.com/a/plvlGg7
Now adding just a single data point of Switzerland increases the growth rate to 34%: https://imgur.com/a/B4zjpzA
This seems wrong. Instead I'd expect that countries with a single data point should not influence the result at all.

**Comment by nicebread** [comment]
Yeah, it's a single regression line through all data points. When mixing different countries, this doesn't make necessarily sense. Probably should be a multilevel model/ different intercepts for each country. But then we also need multiple prediction lines; I have to check whether this clutters the plot too much. Probably as an option.

**Comment by thiemonagel** [comment]
I'd say plotting a single prediction line is probably fine if all countries share the same b. (Imho, plotting one prediction line per country would only make sense if b is determined per country as well.) Maybe the intercept of the global prediction line could be calculated by averaging the per-country intercepts, weighted by number of data points of that country minus 1 (to ignore single data point countries)?

**Comment by khturner** [comment]
You'd probably want to fit it with a mixed model with `country` as a random effect, like `log(cases) ~ date + (1 | country)`. That'd allow for independent estimation of the intercept while fitting the slope together. If you wanted different intercepts and slopes for each country, that'd be `log(cases) ~ country + date:country` (I think).

**Comment by nicebread** [comment]
@khturner: That's what I thought too, but fitting the logarithmized linear model might misestimate a lot (see https://stackoverflow.com/questions/31851936/exponential-curve-fitting-in-r#comment105713296_31854405) - it definitely gives different results compared to the nls estimation. Have to do some research what the best approach would be. (Bayesian modelling takes too much time for the app).

**Comment by khturner** [comment]
Ohhh cool, thanks! Some quick googling points me to the `nlme` package if you want to add random effects for country/state to your nonlinear model, though I've never used it personally:
https://cran.r-project.org/web/packages/nlme/nlme.pdf

**Comment by fknauf** [comment]
@nicebread:

> That's what I thought too, but fitting the logarithmized linear model might misestimate a lot (see https://stackoverflow.com/questions/31851936/exponential-curve-fitting-in-r#comment105713296_31854405) 

I think the stackoverflow answer is mistaken. It's not describing an exponential fit but a polynomial one -- time^n instead of (1+b)^time -- and this leads to strange results. Not only can the linear fit through logarithms not work as well in that setting, but it also means that the sampling error distribution is not such that an LSE loss function over logarithms is really justified.

Here we are guessing that the fit is essentially exponential, and it seems sensible to assume that the sampling error variance increases with the number of cases. I would argue that

cases = (1 + b)^(time + noise)

or equivalently

log(cases) = log(1 + b) * (time + noise)

with homoscedastic noise is a justifiable simple model. In this setting and if we assume normally distributed noise, an LSE loss over f(time) = log(cases) is appropriate. The current LSE loss over the number of cases implicitly assumes that the sampling error is directly normally distributed, i.e.

cases = (1 + b)^time + noise

which seems to me like a bold assumption: it means that an error of 400 cases is equally likely when there are 20000 cases as when there are 100, and to my mind it explains fits like this one (note: logscale y-axis):

![suboptimalfit](https://user-images.githubusercontent.com/56673921/77701806-6278be80-6fb7-11ea-9696-e84f39d6f823.png)

in which the number of cases is overestimated by a factor of 5 in the early stages but suspiciously accurate in the tail end.

So I believe that a linear fit over log(cases) would yield better results, and if I understand him correctly, it would enable @khturner's suggestion. It ultimately comes down to what you believe is a better approximation of the noise on the data, though.

**Comment by nicebread** [comment]
Yes, this is convincing. I'm working on a mixed effects model solution.

**Comment by thiemonagel** [comment]
I think by now the fit has a more fundamental problem: As most countries have implemented physical distancing measures at some point in the past weeks the assumption of a single growth rate over the whole time period is manifestly wrong. Maybe a model should be fitted that includes a time dependency in the exponent?

**Comment by nicebread** [comment]
@thiemonagel: Absolutely. But we should be aware of the scope of this tool - I think it's not possible to transform it to a full-blown modeling tool; it will always fail as such.
The idea is just to provide some rough reference line (and it should be clear to users that this is **not** an actual epidemiological model or prediction. Maybe less is more.

**Comment by fknauf** [comment]
I would be surprised if it were possible to fit such a model from the 25-ish data points we currently have (I'm thinking about a single country) without severe overfitting, especially because the change of the growth rate over time is unlikely to be a very simple function. Moreover, exponential growth is unlikely to be sustained long enough to actually get the amount of data points we'd need to fit such a model. So I agree with @nicebread there, and I expect this to be a big problem for anyone who's trying to build a model for extrapolation -- the future is especially difficult to predict if there are exponential functions involved.

What this can be useful for, I think, is to show the impact of countermeasures after the fact. I have implemented the log-linear fit and obtained interesting results. The new fit on logscale (for comparison with the above) is

![resultingfit](https://user-images.githubusercontent.com/56673921/77765276-4f5d0180-703e-11ea-9975-d9bfe47913ae.png)

Which on a linear scale looks like this:

![resultingfit_nonlog](https://user-images.githubusercontent.com/56673921/77765303-5c79f080-703e-11ea-8c2a-366f44e194ed.png)

So it looks as though the institution of countermeasures around day 20 has an impact. Compare this with the USA, who are still on a roughly exponential trajectory:

![usafit_nonlog](https://user-images.githubusercontent.com/56673921/77765492-9ea33200-703e-11ea-8118-ff20f4af4792.png)

The option to limit the data points that go into the fit turns out to be rather useful in some cases. For example, by limiting the data that goes into the German plot to the first 20 days before the noticeable change in behavior, we get this plot:

![germany_20days](https://user-images.githubusercontent.com/56673921/77765940-41f44700-703f-11ea-8752-2eaf0a92a795.png)

that I find unreasonably sexy. It's probably a coincidence that the values align quite that perfectly.

I can polish this into a pull request if you're interested, although I have to warn you that R is not what you might consider my native programming language.

**Comment by nicebread** [comment]
@fknauf: Maybe take a look at this branch: https://github.com/nicebread/corona/tree/mixed-effects-fit

There I implemented the (definitely preferable) lm approach with `log(cum_cases) ~ days` when one country is selected, and a hierarchical model with random intercepts and random slopes when multiple countries are selected. The plot shows the fixed effect; optionally you can add faint curves with the random slopes:

![grafik](https://user-images.githubusercontent.com/2391270/77780592-83432180-7054-11ea-8dbd-855e75ea6f8f.png)


Any review is appreciated! (And also PRs if you want to tackle this).
My plan is to change the UI, so that you either get the fitted curve, or you can place a manual curve into the plot with the sliders.

**Comment by nicebread** [comment]
New estimation now implemented in master.
